src/combinatorics/simple_graph/regularity/chunk.lean

```diff
@@ -0,0 +1,550 @@
+/-
+Copyright (c) 2022 Yaël Dillies, Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import combinatorics.simple_graph.regularity.bound
+import combinatorics.simple_graph.regularity.equitabilise
+import combinatorics.simple_graph.regularity.uniform
+
+/-!
+# Chunk of the increment partition for Szemerédi Regularity Lemma
+
+In the proof of Szemerédi Regularity Lemma, we need to partition each part of a starting partition
+to increase the energy. This file defines those partitions of parts and shows that they locally
+increase the energy.
+
+This entire file is internal to the proof of Szemerédi Regularity Lemma.
+
+## Main declarations
+
+* `szemeredi_regularity.chunk`: The partition of a part of the starting partition.
+* `szemeredi_regularity.edge_density_chunk_uniform`: `chunk` does not locally decrease the edge
+  density between uniform parts too much.
+* `szemeredi_regularity.edge_density_chunk_not_uniform`: `chunk` locally increases the edge density
+  between non-uniform parts.
+
+## TODO
+
+Once ported to mathlib4, this file will be a great golfing ground for Heather's new tactic
+`rel_congr`.
+-/
+
+open finpartition finset fintype rel nat
+open_locale big_operators classical
+
+local attribute [positivity] tactic.positivity_szemeredi_regularity
+
+namespace szemeredi_regularity
+variables {α : Type*} [fintype α] {P : finpartition (univ : finset α)} (hP : P.is_equipartition)
+  (G : simple_graph α) (ε : ℝ) {U : finset α} (hU : U ∈ P.parts) (V : finset α)
+
+local notation `m` := (card α/step_bound P.parts.card : ℕ)
+
+/-!
+### Definitions
+
+We define `chunk`, the partition of a part, and `star`, the sets of parts of `chunk` that are
+contained in the corresponding witness of non-uniformity.
+-/
+
+/-- The portion of `szemeredi_regularity.increment` which partitions `U`. -/
+noncomputable def chunk : finpartition U :=
+dite (U.card = m * 4 ^ P.parts.card + (card α/P.parts.card - m * 4 ^ P.parts.card))
+  (λ hUcard, (atomise U $ P.nonuniform_witnesses G ε U).equitabilise $ card_aux₁ hUcard)
+  (λ hUcard, (atomise U $ P.nonuniform_witnesses G ε U).equitabilise $ card_aux₂ hP hU hUcard)
```

Comment:
```suggestion
if hUcard : U.card = m * 4 ^ P.parts.card + (card α/P.parts.card - m * 4 ^ P.parts.card
  then (atomise U $ P.nonuniform_witnesses G ε U).equitabilise $ card_aux₁ hUcard)
  else (atomise U $ P.nonuniform_witnesses G ε U).equitabilise $ card_aux₂ hP hU hUcard)
```

Reply:
Can the `if` be moved inside `(atomise U $ P.nonuniform_witnesses G ε U).equitabilise`?

Reply:
No, see the comment below